File: btrfs-hash.c
#include "crypto/xxhash.h"
#include "kerncompat.h"
#include "kernel-shared/accessors.h"
#include "kernel-shared/ctree.h"
#include "kernel-shared/disk-io.h"
#include "kernel-shared/extent_io.h"
#include "kernel-shared/uapi/btrfs_tree.h"
#include "kernel-shared/volumes.h"
#include <blkid/blkid.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

static int btrfs_lookup_csums(
	struct btrfs_trans_handle* trans,
	struct btrfs_root* root,
	struct btrfs_path* path,
	u64 bytenr,
	int total_csums,
	XXH64_hash_t* xxhash
) {
	u16 csum_size = root->fs_info->csum_size;
	int pending_csums = total_csums;
	u8 csum_buf[64]; // Btrfs supports up to 64 bytes (BLAKE2b)

	struct btrfs_key file_key;
	file_key.objectid = BTRFS_EXTENT_CSUM_OBJECTID;
	file_key.offset = bytenr;
	file_key.type = BTRFS_EXTENT_CSUM_KEY;

	int ret = btrfs_search_slot(trans, root, &file_key, path, 0, 0);
	if (ret < 0) goto fail;
	if (ret > 0) {
		if (path->slots[0] == 0) {
			ret = btrfs_prev_leaf(root, path);
			if (ret < 0) goto fail;
			if (ret > 0) {
				ret = -ENOENT;
				goto fail;
			}
		} else {
			path->slots[0]--;
		}
	}

	while (pending_csums > 0) {
		struct extent_buffer* leaf = path->nodes[0];
		int slot = path->slots[0];
		if (slot >= btrfs_header_nritems(leaf)) {
			ret = btrfs_next_leaf(root, path);
			if (ret < 0) goto fail;
			if (ret > 0) break;
			continue;
		}

		struct btrfs_key found_key;
		btrfs_item_key_to_cpu(leaf, &found_key, slot);
		if (found_key.objectid != BTRFS_EXTENT_CSUM_OBJECTID ||
		    found_key.type != BTRFS_EXTENT_CSUM_KEY) {
			break;
		}

		u64 item_start = found_key.offset;
		u32 item_size = btrfs_item_size(leaf, slot);
		u64 item_end = item_start + (item_size / csum_size) * root->fs_info->sectorsize;

		if (bytenr >= item_end || bytenr < item_start) {
			if (bytenr < item_start) break;
			path->slots[0]++;
			continue;
		}

		u32 offset_in_item = (bytenr - item_start) / root->fs_info->sectorsize;
		u32 csums_available = (item_size / csum_size) - offset_in_item;
		u32 csums_to_read = (csums_available > pending_csums) ? pending_csums : csums_available;

		struct btrfs_csum_item* item = btrfs_item_ptr(leaf, slot, struct btrfs_csum_item);
		unsigned long ptr = (unsigned long)item + (offset_in_item * csum_size);
		u32 bytes_to_read = csums_to_read * csum_size;

		void* bulk_csums = malloc(bytes_to_read);
		if (bulk_csums) {
			read_extent_buffer(leaf, bulk_csums, ptr, bytes_to_read);
			*xxhash = XXH64(bulk_csums, bytes_to_read, *xxhash);
			free(bulk_csums);
		} else {
			for (u32 i = 0; i < csums_to_read; i++) {
				read_extent_buffer(leaf, csum_buf, ptr, csum_size);
				*xxhash = XXH64(csum_buf, csum_size, *xxhash);
				ptr += csum_size;
			}
		}

		pending_csums -= csums_to_read;
		bytenr += (u64)csums_to_read * root->fs_info->sectorsize;
		path->slots[0]++;
	}

	return (pending_csums == 0) ? 0 : -ENOENT;

fail:
	return (ret < 0) ? ret : -1;
}

static int btrfs_lookup_extent(
	struct btrfs_fs_info* device_info,
	u64 ino,
	u64 subvolid,
	XXH64_hash_t* xxhash
) {
	int total_csums = 0;
	u32 sectorsize = device_info->sectorsize;

	struct btrfs_root* fs_root;
	if (subvolid == BTRFS_FS_TREE_OBJECTID) {
		fs_root = device_info->fs_root;
	} else {
		struct btrfs_key root_key;
		root_key.objectid = subvolid;
		root_key.type = BTRFS_ROOT_ITEM_KEY;
		root_key.offset = (u64)-1;
		fs_root = btrfs_read_fs_root(device_info, &root_key);
		if (IS_ERR(fs_root)) {
			fprintf(stderr, "Error: Unable to read subvolume %llu\n", (unsigned long long)subvolid);
			return PTR_ERR(fs_root);
		}
	}
	struct btrfs_key key;
	key.objectid = ino;
	key.type = BTRFS_EXTENT_DATA_KEY;
	key.offset = 0;
	struct btrfs_key found_key;
	struct extent_buffer* leaf;
	int slot;
	struct btrfs_path* path1 = btrfs_alloc_path();
	struct btrfs_path *path = btrfs_alloc_path();
	int ret = btrfs_search_slot(NULL, fs_root, &key, path, 0, 0);

	if (ret < 0) {
		fprintf(stderr,
			"Error: btrfs_search_slot failed with %d\n", ret);
		goto error;
	}

	if (ret > 0) {
		leaf = path->nodes[0];
		slot = path->slots[0];
		if (slot >= btrfs_header_nritems(leaf)) {
			ret = btrfs_next_leaf(fs_root, path);
			if (ret < 0) goto error;
			if (ret > 0) {
				ret = -ENOENT;
				goto error;
			}
			leaf = path->nodes[0];
			slot = path->slots[0];
		}
		btrfs_item_key_to_cpu(leaf, &found_key, slot);

		if (found_key.objectid != ino || found_key.type != BTRFS_EXTENT_DATA_KEY) {
			ret = -ENOENT;
			goto error;
		}
	}

	struct btrfs_root* csum_root = btrfs_csum_root(device_info, 0);

	while (1) {
		leaf = path->nodes[0];
		slot = path->slots[0];
		if (slot >= btrfs_header_nritems(leaf)) {
			ret = btrfs_next_leaf(fs_root, path);
			if (ret == 0) continue;
			if (ret < 0) goto error;
			break;
		}

		btrfs_item_key_to_cpu(leaf, &found_key, slot);
		if (found_key.objectid != ino || found_key.type != BTRFS_EXTENT_DATA_KEY) {
			break;
		}

		struct btrfs_file_extent_item* fi = btrfs_item_ptr(leaf, slot,
					struct btrfs_file_extent_item);
		if (btrfs_file_extent_type(leaf, fi) ==
			BTRFS_FILE_EXTENT_INLINE) {
			ret = -1; // Fallback to direct read
			goto error;
		}

		u64 disk_bytenr = btrfs_file_extent_disk_bytenr(leaf, fi);
		if (disk_bytenr == 0) { // Sparse hole
			path->slots[0]++;
			continue;
		}
		u64 disk_num_bytes = btrfs_file_extent_disk_num_bytes(leaf, fi);
		u64 offset = btrfs_file_extent_offset(leaf, fi);
		u64 num_bytes = btrfs_file_extent_num_bytes(leaf, fi);
		u8 compression = btrfs_file_extent_compression(leaf, fi);
		u64 bytenr;

		if (compression) {
			// Compressed: csums cover the compressed on-disk data
			bytenr = disk_bytenr;
			total_csums = disk_num_bytes / sectorsize;
		} else {
			// Uncompressed: csums cover the referenced portion
			bytenr = disk_bytenr + offset;
			total_csums = num_bytes / sectorsize;
		}

		csum_root = btrfs_csum_root(device_info, bytenr);
		ret = btrfs_lookup_csums(
			NULL,
			csum_root,
			path1,
			bytenr,
			total_csums,
			xxhash
		);
		btrfs_release_path(path1);
		if (ret) goto error;
		path->slots[0]++;
	}
	ret = 0;

error:
	btrfs_free_path(path1);
	btrfs_release_path(path);
	btrfs_free_path(path);
	return ret;
}

static char* resolve_device(const char* file_path, u64* subvolid) {
	*subvolid = BTRFS_FS_TREE_OBJECTID;
	char resolved_file_path[PATH_MAX];
	if (!realpath(file_path, resolved_file_path)) {
		fprintf(stderr, "Error: realpath failed for %s: %s\n", file_path, strerror(errno));
		return NULL;
	}

	FILE* mntinfo = fopen("/proc/self/mountinfo", "r");
	if (!mntinfo) {
		fprintf(stderr, "Error: Unable to open /proc/self/mountinfo\n");
		return NULL;
	}

	// find longest mount_path that is prefix of file_path
	size_t mount_path_length_max = 0;
	static char device_path[4096];
	device_path[0] = '\0';
	// read chunk-wise
	char chunk[4096];
	while (fgets(chunk, 4096, mntinfo)) {
		// index of " - ", right before the fs type
		char* result_address = strstr(chunk, " - btrfs ");
		if (!result_address) continue;
		result_address += 3;
		// check if next comes "btrfs "
		if (*result_address != 'b') continue;
		if (*(++result_address) != 't') continue;
		if (*(++result_address) != 'r') continue;
		if (*(++result_address) != 'f') continue;
		if (*(++result_address) != 's') continue;
		if (*(++result_address) != ' ') continue;

		// skip first 4 columns
		char* mount_path = strtok(chunk, " ");
		int column_count = 0;
		while (mount_path) {
			if (++column_count == 5) goto column_found;
			mount_path = strtok(NULL, " ");
		}
		continue;
		column_found:;

		// check if file_path begins with mount_path
		size_t mount_path_length = strlen(mount_path);
		if (mount_path_length > 1) {
			mount_path[mount_path_length] = '/';
			mount_path[++mount_path_length] = '\0';
		}
		if (
			mount_path_length < mount_path_length_max ||
			strncmp(resolved_file_path, mount_path, mount_path_length)
		) continue;

		char* result_end_address = strstr(++result_address, " ");
		if (!result_end_address) continue;
		*result_end_address = '\0';
		mount_path_length_max = mount_path_length;
		strncpy(device_path, result_address, 4095);

		// parse subvolid from mount options after device path
		char* svid = strstr(++result_end_address, "subvolid=");
		if (svid) {
			*subvolid = (u64)strtoull(svid + 9, NULL, 10);
		}
	}
	if (mount_path_length_max) return device_path;

	fprintf(stderr, "Error: Unable to determine block device for %s\n", resolved_file_path);
	return NULL;
}

int main(int argc, char** argv) {
	if (argc != 2) {
		fprintf(stderr, "Usage: btrfs-hash somehugefile.bin\n");
		return -1;
	}

	char* file_path = argv[1];
	struct stat file_stat;
	if (stat(file_path, &file_stat) < 0) {
		fprintf(stderr, "Error: Unable to access file %s: %s\n", file_path, strerror(errno));
		return -1;
	}
	if (!S_ISREG(file_stat.st_mode)) {
		fprintf(stderr, "Error: %s is not a regular file.\n", file_path);
		return -1;
	}

	u64 subvolid;
	char* device_path = resolve_device(file_path, &subvolid);
	if (!device_path) return -1;
	struct open_ctree_args open_ctree_fs_info_args = {
		.flags = OPEN_CTREE_PARTIAL,
		.filename = device_path,
	};
	struct btrfs_fs_info* device_info = open_ctree_fs_info(&open_ctree_fs_info_args);
	if (!device_info) {
		fprintf(stderr, "Error: Unable to open block device %s\n", device_path);
		return -1;
	}

	XXH64_hash_t xxhash = 0;
	if (
		btrfs_lookup_extent(
			device_info,
			file_stat.st_ino,
			subvolid,
			&xxhash
		)
	) { // if lookup fails, hash the file directly
		fprintf(stderr, "Unable to lookup extent, falling back to direct read...\n");
		int file = open(file_path, O_RDONLY);
		if (file < 0) {
			fprintf(stderr, "Error: Unable to open file %s: %s\n", file_path, strerror(errno));
			return -1;
		}

		unsigned char* buffer = malloc(65536);
		if (!buffer) {
			fprintf(stderr, "Error: Unable to allocate buffer\n");
			return -1;
		}
		ssize_t bytes_read;
		while ((bytes_read = read(file, buffer, 65536)) > 0) {
			xxhash = XXH64(buffer, bytes_read, xxhash);
		}
		close(file);
		free(buffer);
		if (bytes_read < 0) {
			fprintf(stderr, "Error: Unable to read file %s: %s\n", file_path, strerror(errno));
			return -1;
		}
	}

	printf("%016llx\n", (unsigned long long)xxhash);
}

File: btrfs-hash.zig
const std = @import("std");

const stderr = std.io.getStdErr().writer();

// Btrfs types are opaque to Zig (anonymous unions/structs in headers).
// Field access and inline functions are provided by shim.c.
const BtrfsFsInfo = opaque {};
const BtrfsRoot = opaque {};
const BtrfsPath = opaque {};
const ExtentBuffer = opaque {};

const BtrfsKey = extern struct {
	objectid: u64,
	type: u8,
	offset: u64 align(1),
};

const OpenCtreeArgs = extern struct {
	filename: ?[*:0]const u8 = null,
	sb_bytenr: u64 = 0,
	root_tree_bytenr: u64 = 0,
	chunk_tree_bytenr: u64 = 0,
	flags: c_uint = 0,
};

// btrfs_tree.h constants
const BTRFS_EXTENT_CSUM_OBJECTID: u64 = @bitCast(@as(i64, -10));
const BTRFS_FS_TREE_OBJECTID: u64 = 5;
const BTRFS_EXTENT_CSUM_KEY: u8 = 128;
const BTRFS_EXTENT_DATA_KEY: u8 = 108;
const BTRFS_ROOT_ITEM_KEY: u8 = 132;
const BTRFS_FILE_EXTENT_INLINE: u8 = 0;
const OPEN_CTREE_PARTIAL: c_uint = 1 << 1;

// btrfs-progs C functions
extern fn open_ctree_fs_info(oca: *OpenCtreeArgs) ?*BtrfsFsInfo;
extern fn btrfs_read_fs_root(info: *BtrfsFsInfo, key: *BtrfsKey) ?*BtrfsRoot;
extern fn btrfs_csum_root(info: *BtrfsFsInfo, bytenr: u64) ?*BtrfsRoot;
extern fn btrfs_alloc_path() ?*BtrfsPath;
extern fn btrfs_free_path(path: *BtrfsPath) void;
extern fn btrfs_release_path(path: *BtrfsPath) void;
extern fn btrfs_search_slot(
	trans: ?*anyopaque, root: *BtrfsRoot,
	key: *BtrfsKey, path: *BtrfsPath,
	ins_len: c_int, cow: c_int,
) c_int;
extern fn btrfs_prev_leaf(root: *BtrfsRoot, path: *BtrfsPath) c_int;
extern fn read_extent_buffer(
	eb: *const ExtentBuffer, dst: ?*anyopaque,
	start: c_ulong, len: c_ulong,
) void;
extern fn XXH64(input: ?*const anyopaque, length: usize, seed: u64) u64;

// shim.c wrappers for inline/macro functions and opaque struct fields
extern fn shim_fs_info_sectorsize(info: *const BtrfsFsInfo) u32;
extern fn shim_fs_info_csum_size(info: *const BtrfsFsInfo) u16;
extern fn shim_fs_info_fs_root(info: *const BtrfsFsInfo) ?*BtrfsRoot;
extern fn shim_root_fs_info(root: *const BtrfsRoot) *BtrfsFsInfo;
extern fn shim_path_node(path: *const BtrfsPath, level: c_int) ?*ExtentBuffer;
extern fn shim_path_slot(path: *const BtrfsPath, level: c_int) c_int;
extern fn shim_path_inc_slot(path: *BtrfsPath, level: c_int) void;
extern fn shim_path_dec_slot(path: *BtrfsPath, level: c_int) void;
extern fn shim_header_nritems(eb: *const ExtentBuffer) u32;
extern fn shim_item_key_to_cpu(eb: *const ExtentBuffer, key: *BtrfsKey, nr: c_int) void;
extern fn shim_next_leaf(root: *BtrfsRoot, path: *BtrfsPath) c_int;
extern fn shim_item_size(eb: *const ExtentBuffer, slot: c_int) u32;
extern fn shim_item_ptr_offset(leaf: *ExtentBuffer, slot: c_int) c_ulong;
extern fn shim_file_extent_type(eb: *ExtentBuffer, fi: c_ulong) u8;
extern fn shim_file_extent_disk_bytenr(eb: *ExtentBuffer, fi: c_ulong) u64;
extern fn shim_file_extent_disk_num_bytes(eb: *ExtentBuffer, fi: c_ulong) u64;
extern fn shim_file_extent_offset(eb: *ExtentBuffer, fi: c_ulong) u64;
extern fn shim_file_extent_num_bytes(eb: *ExtentBuffer, fi: c_ulong) u64;
extern fn shim_file_extent_compression(eb: *ExtentBuffer, fi: c_ulong) u8;

fn isErrPtr(ptr: ?*anyopaque) bool {
	return @intFromPtr(ptr) >= @as(usize, 0) -% 4095;
}

fn lookupCsums(
	root: *BtrfsRoot,
	path: *BtrfsPath,
	start_bytenr: u64,
	total_csums: u32,
	xxhash: *u64,
) bool {
	const info = shim_root_fs_info(root);
	const csum_size: u32 = shim_fs_info_csum_size(info);
	const sectorsize: u64 = shim_fs_info_sectorsize(info);
	var pending: u32 = total_csums;
	var bytenr = start_bytenr;

	var key = BtrfsKey{
		.objectid = BTRFS_EXTENT_CSUM_OBJECTID,
		.type = BTRFS_EXTENT_CSUM_KEY,
		.offset = bytenr,
	};

	var ret = btrfs_search_slot(null, root, &key, path, 0, 0);
	if (ret < 0) return false;
	if (ret > 0) {
		if (shim_path_slot(path, 0) == 0) {
			ret = btrfs_prev_leaf(root, path);
			if (ret != 0) return false;
		} else {
			shim_path_dec_slot(path, 0);
		}
	}

	while (pending > 0) {
		const leaf = shim_path_node(path, 0) orelse return false;
		const slot = shim_path_slot(path, 0);
		if (slot >= @as(c_int, @intCast(shim_header_nritems(leaf)))) {
			ret = shim_next_leaf(root, path);
			if (ret < 0) return false;
			if (ret > 0) break;
			continue;
		}

		var found_key: BtrfsKey = undefined;
		shim_item_key_to_cpu(leaf, &found_key, slot);
		if (found_key.objectid != BTRFS_EXTENT_CSUM_OBJECTID or
			found_key.type != BTRFS_EXTENT_CSUM_KEY)
			break;

		const item_start = found_key.offset;
		const item_size = shim_item_size(leaf, slot);
		const item_end = item_start + (item_size / csum_size) * sectorsize;
		if (bytenr >= item_end or bytenr < item_start) {
			if (bytenr < item_start) break;
			shim_path_inc_slot(path, 0);
			continue;
		}

		const offset_in_item: u32 = @intCast((bytenr - item_start) / sectorsize);
		const csums_available = (item_size / csum_size) - offset_in_item;
		const csums_to_read = @min(csums_available, pending);
		const bytes_to_read = csums_to_read * csum_size;
		const ptr = shim_item_ptr_offset(leaf, slot) + offset_in_item * csum_size;

		var buf: [65536]u8 = undefined;
		if (bytes_to_read <= buf.len) {
			read_extent_buffer(leaf, &buf, ptr, bytes_to_read);
			xxhash.* = XXH64(&buf, bytes_to_read, xxhash.*);
		} else {
			// fallback: read csum by csum
			var p = ptr;
			var csum_buf: [64]u8 = undefined;
			for (0..csums_to_read) |_| {
				read_extent_buffer(leaf, &csum_buf, p, csum_size);
				xxhash.* = XXH64(&csum_buf, csum_size, xxhash.*);
				p += csum_size;
			}
		}

		pending -= csums_to_read;
		bytenr += @as(u64, csums_to_read) * sectorsize;
		shim_path_inc_slot(path, 0);
	}

	return pending == 0;
}

fn lookupExtent(
	device_info: *BtrfsFsInfo,
	ino: u64,
	subvolid: u64,
	xxhash: *u64,
) bool {
	const sectorsize: u64 = shim_fs_info_sectorsize(device_info);

	// resolve fs root for the subvolume
	const fs_root: *BtrfsRoot = blk: {
		if (subvolid == BTRFS_FS_TREE_OBJECTID)
			break :blk shim_fs_info_fs_root(device_info) orelse return false;
		var root_key = BtrfsKey{
			.objectid = subvolid,
			.type = BTRFS_ROOT_ITEM_KEY,
			.offset = @bitCast(@as(i64, -1)),
		};
		const result = btrfs_read_fs_root(device_info, &root_key);
		if (isErrPtr(@ptrCast(result))) {
			stderr.print("Error: Unable to read subvolume {d}\n", .{subvolid}) catch {};
			return false;
		}
		break :blk result orelse return false;
	};

	var key = BtrfsKey{
		.objectid = ino,
		.type = BTRFS_EXTENT_DATA_KEY,
		.offset = 0,
	};

	const path1 = btrfs_alloc_path() orelse return false;
	defer btrfs_free_path(path1);
	const path = btrfs_alloc_path() orelse return false;
	defer {
		btrfs_release_path(path);
		btrfs_free_path(path);
	}

	var ret = btrfs_search_slot(null, fs_root, &key, path, 0, 0);
	if (ret < 0) {
		stderr.print("Error: btrfs_search_slot failed with {d}\n", .{ret}) catch {};
		return false;
	}

	if (ret > 0) {
		var leaf = shim_path_node(path, 0) orelse return false;
		var slot = shim_path_slot(path, 0);
		if (slot >= @as(c_int, @intCast(shim_header_nritems(leaf)))) {
			ret = shim_next_leaf(fs_root, path);
			if (ret != 0) return false;
			leaf = shim_path_node(path, 0) orelse return false;
			slot = shim_path_slot(path, 0);
		}
		var found_key: BtrfsKey = undefined;
		shim_item_key_to_cpu(leaf, &found_key, slot);
		if (found_key.objectid != ino or found_key.type != BTRFS_EXTENT_DATA_KEY)
			return false;
	}

	// iterate file extents
	while (true) {
		const leaf = shim_path_node(path, 0) orelse return false;
		const slot = shim_path_slot(path, 0);
		if (slot >= @as(c_int, @intCast(shim_header_nritems(leaf)))) {
			ret = shim_next_leaf(fs_root, path);
			if (ret == 0) continue;
			if (ret < 0) return false;
			break;
		}

		var found_key: BtrfsKey = undefined;
		shim_item_key_to_cpu(leaf, &found_key, slot);
		if (found_key.objectid != ino or found_key.type != BTRFS_EXTENT_DATA_KEY)
			break;

		const fi = shim_item_ptr_offset(leaf, slot);
		if (shim_file_extent_type(leaf, fi) == BTRFS_FILE_EXTENT_INLINE)
			return false; // fallback to direct read

		const disk_bytenr = shim_file_extent_disk_bytenr(leaf, fi);
		if (disk_bytenr == 0) { // sparse hole
			shim_path_inc_slot(path, 0);
			continue;
		}

		const disk_num_bytes = shim_file_extent_disk_num_bytes(leaf, fi);
		const extent_offset = shim_file_extent_offset(leaf, fi);
		const num_bytes = shim_file_extent_num_bytes(leaf, fi);
		const compression = shim_file_extent_compression(leaf, fi);

		// compressed: csums cover the on-disk data
		// uncompressed: csums cover the referenced portion
		const bytenr = if (compression != 0) disk_bytenr else disk_bytenr + extent_offset;
		const total_csums: u32 = @intCast(
			(if (compression != 0) disk_num_bytes else num_bytes) / sectorsize,
		);

		const csum_root = btrfs_csum_root(device_info, bytenr) orelse return false;
		if (!lookupCsums(csum_root, path1, bytenr, total_csums, xxhash))
			return false;
		btrfs_release_path(path1);
		shim_path_inc_slot(path, 0);
	}

	return true;
}

const ResolveResult = struct {
	device_path: []const u8,
	subvolid: u64,
};

fn resolveDevice(file_path: [*:0]const u8) ?ResolveResult {
	var resolved_buf: [std.posix.PATH_MAX]u8 = undefined;
	const resolved = std.posix.realpathZ(file_path, &resolved_buf) catch {
		stderr.print("Error: realpath failed for {s}\n", .{file_path}) catch {};
		return null;
	};

	const mntinfo = std.fs.openFileAbsolute("/proc/self/mountinfo", .{}) catch {
		stderr.print("Error: Unable to open /proc/self/mountinfo\n", .{}) catch {};
		return null;
	};
	defer mntinfo.close();

	const S = struct {
		var device_path_buf: [4096]u8 = undefined;
		var device_path_len: usize = 0;
	};

	var best_prefix_len: usize = 0;
	var subvolid: u64 = BTRFS_FS_TREE_OBJECTID;
	var line_buf: [8192]u8 = undefined;
	const reader = mntinfo.reader();

	while (reader.readUntilDelimiter(&line_buf, '\n')) |line| {
		// find " - btrfs " marker, right before the fs type
		const marker = " - btrfs ";
		const marker_pos = std.mem.indexOf(u8, line, marker) orelse continue;
		const after_marker = line[marker_pos + marker.len ..];

		// extract mount_path (5th column)
		var col_iter = std.mem.tokenizeScalar(u8, line[0..marker_pos], ' ');
		var col: usize = 0;
		var mount_path: []const u8 = "";
		while (col_iter.next()) |token| {
			col += 1;
			if (col == 5) {
				mount_path = token;
				break;
			}
		}
		if (col < 5) continue;

		// check if file_path begins with mount_path
		var mp_len = mount_path.len;
		if (mp_len > 1) {
			if (!std.mem.startsWith(u8, resolved, mount_path)) continue;
			if (resolved.len > mp_len and resolved[mp_len] != '/') continue;
			mp_len += 1;
		} else {
			if (!std.mem.startsWith(u8, resolved, "/")) continue;
		}
		if (mp_len < best_prefix_len) continue;

		// extract device path (first token after "btrfs ")
		const dev_end = std.mem.indexOfScalar(u8, after_marker, ' ') orelse after_marker.len;
		const dev_path = after_marker[0..dev_end];
		@memcpy(S.device_path_buf[0..dev_path.len], dev_path);
		S.device_path_len = dev_path.len;
		best_prefix_len = mp_len;

		// parse subvolid from mount options
		const opts = if (dev_end < after_marker.len) after_marker[dev_end..] else "";
		if (std.mem.indexOf(u8, opts, "subvolid=")) |svid_pos| {
			const num_start = svid_pos + "subvolid=".len;
			var num_end = num_start;
			while (num_end < opts.len and opts[num_end] >= '0' and opts[num_end] <= '9')
				num_end += 1;
			subvolid = std.fmt.parseUnsigned(u64, opts[num_start..num_end], 10) catch
				BTRFS_FS_TREE_OBJECTID;
		}
	} else |_| {}

	if (best_prefix_len == 0) {
		stderr.print("Error: Unable to determine block device for {s}\n", .{resolved}) catch {};
		return null;
	}

	S.device_path_buf[S.device_path_len] = 0; // null-terminate for C
	return .{
		.device_path = S.device_path_buf[0..S.device_path_len],
		.subvolid = subvolid,
	};
}

fn directHash(file_path: [*:0]const u8) ?u64 {
	const file = std.fs.cwd().openFileZ(file_path, .{}) catch |err| {
		stderr.print("Error: Unable to open file {s}: {}\n", .{ file_path, err }) catch {};
		return null;
	};
	defer file.close();

	var xxhash: u64 = 0;
	var buf: [65536]u8 = undefined;
	while (true) {
		const n = file.read(&buf) catch |err| {
			stderr.print("Error: Unable to read file {s}: {}\n", .{ file_path, err }) catch {};
			return null;
		};
		if (n == 0) break;
		xxhash = XXH64(&buf, n, xxhash);
	}
	return xxhash;
}

pub fn main() !void {
	const args = try std.process.argsAlloc(std.heap.c_allocator);
	defer std.process.argsFree(std.heap.c_allocator, args);

	if (args.len != 2) {
		try stderr.print("Usage: btrfs-hash somehugefile.bin\n", .{});
		std.process.exit(255);
	}

	const file_path: [*:0]const u8 = args[1];
	const stat = std.posix.fstatat(std.posix.AT.FDCWD, std.mem.span(file_path), 0) catch {
		try stderr.print("Error: Unable to access file {s}\n", .{file_path});
		std.process.exit(255);
	};
	if (stat.mode & std.posix.S.IFMT != std.posix.S.IFREG) {
		try stderr.print("Error: {s} is not a regular file.\n", .{file_path});
		std.process.exit(255);
	}

	const resolve = resolveDevice(file_path) orelse std.process.exit(255);

	var oca = OpenCtreeArgs{
		.flags = OPEN_CTREE_PARTIAL,
		.filename = @ptrCast(resolve.device_path.ptr),
	};
	const device_info = open_ctree_fs_info(&oca) orelse {
		try stderr.print("Error: Unable to open block device {s}\n", .{resolve.device_path});
		std.process.exit(255);
	};

	var xxhash: u64 = 0;
	if (!lookupExtent(device_info, stat.ino, resolve.subvolid, &xxhash)) {
		stderr.print("Unable to lookup extent, falling back to direct read...\n", .{}) catch {};
		xxhash = directHash(file_path) orelse std.process.exit(255);
	}

	const stdout = std.io.getStdOut().writer();
	try stdout.print("{x:0>16}\n", .{xxhash});
}
